server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java

```diff
@@ -272,6 +296,64 @@ public void onFailure(Exception e) {
+                                                                                        Set<String> autoCreateIndices) {
+        Map<String, IndexTemplateV2.DataStreamTemplate> autoCreateDataStreams = new HashMap<>();
+        Iterator<String> autoCreateIndicesIterator = autoCreateIndices.iterator();
+        while (autoCreateIndicesIterator.hasNext()) {
+            String indexName = autoCreateIndicesIterator.next();
+            String v2Template = MetadataIndexTemplateService.findV2Template(metadata, indexName, false);
+            if (v2Template != null) {
+                IndexTemplateV2 indexTemplateV2 = metadata.templatesV2().get(v2Template);
+                if (indexTemplateV2.getDataStreamTemplate() != null) {
+                    autoCreateIndicesIterator.remove();
+                    autoCreateDataStreams.put(indexName, indexTemplateV2.getDataStreamTemplate());
+                }
+            }
+        }
+        return autoCreateDataStreams;
+    }
+
+    void autoCreateDataStreams(final Task task,
+                               final BulkRequest bulkRequest,
+                               final long startTimeNanos,
+                               final ActionListener<BulkResponse> listener,
+                               final AtomicArray<BulkItemResponse> responses,
+                               final Map<String, IndexTemplateV2.DataStreamTemplate> autoCreateDataStreams,
+                               final Map<String, IndexNotFoundException> indicesThatCannotBeCreated) {
+        final AtomicInteger counter = new AtomicInteger(autoCreateDataStreams.size());
+        for (Map.Entry<String, IndexTemplateV2.DataStreamTemplate> entry : autoCreateDataStreams.entrySet()) {
+            final String name = entry.getKey();
+
+            CreateDataStreamAction.Request request = new CreateDataStreamAction.Request(name);
+            request.setTimestampFieldName(entry.getValue().getTimestampField());
+            CheckedConsumer<AcknowledgedResponse, ? extends Exception> onResponse = response -> {
+                if (counter.decrementAndGet() == 0) {
+                    threadPool.executor(ThreadPool.Names.WRITE).execute(
+                        () -> executeBulk(task, bulkRequest, startTimeNanos, listener, responses, indicesThatCannotBeCreated));
+                }
+            };
+            Consumer<Exception> onFailure = e -> {
+                if (!(ExceptionsHelper.unwrapCause(e) instanceof ResourceAlreadyExistsException)) {
+                    // fail all requests involving this index, if create didn't work
+                    for (int i = 0; i < bulkRequest.requests.size(); i++) {
+                        DocWriteRequest<?> item = bulkRequest.requests.get(i);
+                        if (item != null && setResponseFailureIfIndexMatches(responses, i, item, name, e)) {
+                            bulkRequest.requests.set(i, null);
+                        }
+                    }
+                }
```

Comment:
Duplicates code in the `onFailure` handler in the `doExecute` method. Might be worth consolidating.